File: exact_root_websocket_server_haskell/src/ExactRoot.hs
module ExactRoot
(
  berechneExacteWurzel,
  Res( .. )
) where
--import Data.ByteString (partition)
import Data.List

data Res = Res {
  multiplikator :: Int,
  wurzelwert :: Int,
  radikand :: Int
} deriving (Eq, Show)

data Res' = Res' {
  multiplikator' :: Int,
  wurzelwert' :: Int,
  radikand' :: Int
} deriving (Eq, Show)

berechneExacteWurzel :: Int -> [Res]
berechneExacteWurzel radikand = do
  let ungeradeZahlen      = berechneUngeradeZahlen radikand
  let einfacheReihe       = berechneEinfacheReihe radikand
  let standardWerte       = reduziereStandardWerte radikand (berechneStandardWerte ungeradeZahlen)
  let radikandWurzelwerte = zippen einfacheReihe standardWerte
  let einfacheWurzelwerte = berechneEinfacheWurzelwert radikand radikandWurzelwerte
  let komplexeWurzelwerte = berechneKomplexeWurzelwert radikand radikandWurzelwerte
  let komplexeWurzelwerte' = berechneKomplexeWurzelwert' radikand radikandWurzelwerte
  -- case einfacheWurzelwerte of
  --  Just w  -> Res (-1) w (-1)
  --  Nothing -> case komplexeWurzelwerte of 
  --      Just (m, w)  -> Res m w (-1)
  --      Nothing   -> Res (-1) (-1) radikand
  case einfacheWurzelwerte of 
    Just w -> [Res (-1) w (-1)]
    Nothing -> map (\(r,w) -> Res r w (-1)) komplexeWurzelwerte' 

berechneUngeradeZahlen :: Int -> [Int]
berechneUngeradeZahlen radikand = filter odd [1 .. radikand]

berechneEinfacheReihe :: Int -> [Int]
berechneEinfacheReihe radikand = [2 .. ((radikand `quot` 2) +1)]

-- die Berechnung bis Resultat der Addition < Radikand 
berechneStandardWerte :: [Int] -> [Int]
berechneStandardWerte  [] = []
berechneStandardWerte  [x] = []
berechneStandardWerte  (x:y:xs) = sum : berechneStandardWerte (sum : xs)
  where sum = x + y

reduziereStandardWerte :: Int -> [Int] -> [Int]
reduziereStandardWerte radiKand stWerte = fst part ++ [head (snd part)]
  where part = partition (< radiKand) stWerte 

zippen :: [Int] -> [Int] -> [(Int, Int)]
zippen = zip

berechneEinfacheWurzelwert :: Int -> [(Int, Int)] -> Maybe Int
berechneEinfacheWurzelwert radikand radikandWurzelwert = auspacken sucheEinfacheWurzelWert
  where
    auspacken :: [(Int, Int)] -> Maybe Int
    auspacken [(r,_)] = Just r
    auspacken ((_,_):y) = Nothing
    auspacken []  = Nothing
    sucheEinfacheWurzelWert :: [(Int, Int)]
    sucheEinfacheWurzelWert = filter (\(r, w) -> w == radikand) radikandWurzelwert

berechneKomplexeWurzelwert :: Int -> [(Int, Int)] -> Maybe (Int, Int)
berechneKomplexeWurzelwert radikand radikandWurzelwert = auspacken sucheKomplexeWurzelwert
  where
    auspacken :: [(Int, Int)] -> Maybe(Int, Int)
    auspacken [] = Nothing
    auspacken [(r, w)] = Just (r, radikand `quot` w)
    auspacken ((r, w) : y) = Just(r, radikand `quot` w)
    sucheKomplexeWurzelwert :: [(Int, Int)]
    sucheKomplexeWurzelwert = filter(\(r, w) -> (radikand `mod` w) == 0) radikandWurzelwert

berechneKomplexeWurzelwert' :: Int -> [(Int, Int)] -> [(Int, Int)]
berechneKomplexeWurzelwert' radikand radikandWurzelwert = auspacken sucheKomplexeWurzelwert
  where
    auspacken :: [(Int, Int)] -> [(Int, Int)]
    auspacken [] = []
    auspacken [(r, w)] = [(r, radikand `quot` w)]
    auspacken ((r, w) : y) = (r, radikand `quot` w): auspacken y
    sucheKomplexeWurzelwert :: [(Int, Int)]
    sucheKomplexeWurzelwert = filter(\(r, w) -> (radikand `mod` w) == 0) radikandWurzelwert

File: exact_root_websocket_server_rust/src/exact_root.rs
pub mod calc_exact_root {
  pub struct Res{
    pub multiplikator : u64,
    pub wurzelwert :u64
  }

  pub fn berechne_exacte_wurzel(radikand : u64) -> Res {
    let ungerade_zahlen_1 = berechne_ungerade_zahlen(radikand);
    let standard_werte_1 = berechne_standard_werte(ungerade_zahlen_1);
    let einfache_reihe_1 = berechne_einfache_reihe(radikand);
    let radikand_wurzelwert_1 = zippen(standard_werte_1, einfache_reihe_1);
    let einfache_wurzelwert = berechne_einfache_wurzelwert(radikand, radikand_wurzelwert_1);
    
    let ungerade_zahlen_2 = berechne_ungerade_zahlen(radikand);
    let standard_werte_2 = berechne_standard_werte(ungerade_zahlen_2);
    let einfache_reihe_2 = berechne_einfache_reihe(radikand);
    let radikand_wurzelwert_2 = zippen(standard_werte_2, einfache_reihe_2);
    let komplexe_wurzelwert = berechne_komplexe_wurzelwert(radikand, radikand_wurzelwert_2);
    let multiplikator = komplexe_wurzelwert.unwrap().0;
    let wurzelwert = komplexe_wurzelwert.unwrap().1;
    match einfache_wurzelwert {
      Some(res) => Res{multiplikator : 1, wurzelwert : res},
      None => Res{multiplikator : multiplikator, wurzelwert : wurzelwert}
    }
    // unberehenbare Wurzel loesen panic aus
  }
  //die Berechnung bis Resultat der Addition < Radikand
  fn berechne_standard_werte(ungerade_zahlen: Vec<u64>) -> Vec<u64> {
    let mut counter = 0;
    let mut result : Vec<u64> = Vec::new();
    let size = ungerade_zahlen.len();
    while counter < (size - 1) {
      if counter == 0 {
        let temp = ungerade_zahlen[counter] + ungerade_zahlen[counter + 1];
        result.push(temp);
      }else{
        let temp = result[counter - 1] + ungerade_zahlen[counter + 1];
        result.push(temp);
      }
      counter += 1;
    }
    result
  }

  use std::vec::Vec;
  fn berechne_ungerade_zahlen(radikand: u64) -> Vec<u64>{
    if radikand < 1 {
      vec![1]
    }else {
      (0..radikand).collect::<Vec<u64>>()
      .into_iter()
      .filter(|x| x % 2 != 0)
      .collect::<Vec<u64>>()
    }
    
  } 

  fn berechne_einfache_reihe(radikand: u64) -> Vec<u64> {
    let teiler = (radikand / 2) + 1;
    (2..teiler).collect::<Vec<u64>>()
  }

  fn zippen(standard_werte: Vec<u64>, einfache_reihe: Vec<u64>) -> Vec<(u64,u64)> {
    let mut result : Vec<(u64, u64)> = Vec::new();
    for (x, y) in einfache_reihe.iter().zip(standard_werte.iter()) {
      result.push((*x, *y));
    }
    result
  }

  fn berechne_einfache_wurzelwert(radikand: u64, radikand_wurzelwert: Vec<(u64, u64)>) -> Option<u64> {
    let result: Vec<(u64, u64)> = radikand_wurzelwert.into_iter()
        .filter(|(_, y)|  *y == radikand).collect::<Vec<(u64, u64)>>();
    if result.len() == 0 {
      None
    }else{
      Some(result[0].0)
    }
  }
      
  fn berechne_komplexe_wurzelwert(radikand: u64, radikand_wurzelwert: Vec<(u64, u64)>) -> Option<(u64, u64)> {
    let mut result: Option<(u64, u64)> = None;
    for (x, y) in radikand_wurzelwert.iter() {
      let temp = radikand % y;
      if temp == 0 {
        result = Some((*x, (radikand / y)));
        break;
      }
    } 
    result
  }
}